src/infra/core/src/ingest/fetch_service/core.rs

```diff
@@ -209,39 +251,15 @@ impl FetchService {
         s: &'a str,
         dataset_env_vars: &'a HashMap<String, DatasetEnvVar>,
     ) -> Result<Cow<'a, str>, PollingIngestError> {
-        let mut s = Cow::from(s);
-        let re_tpl = regex::Regex::new(r"\$\{\{([^}]*)\}\}").unwrap();
-        let re_env = regex::Regex::new(r"^env\.([a-zA-Z-_]+)$").unwrap();
-
-        loop {
-            if let Some(ctpl) = re_tpl.captures(&s) {
-                let tpl_range = ctpl.get(0).unwrap().range();
-
-                if let Some(cenv) = re_env.captures(ctpl.get(1).unwrap().as_str().trim()) {
-                    let env_name = cenv.get(1).unwrap().as_str();
-
-                    let dataset_env_var_secret_value = self
-                        .dataset_key_value_svc
-                        .find_dataset_env_var_value_by_key(env_name, dataset_env_vars)?;
-
-                    s.to_mut()
-                        .replace_range(tpl_range, dataset_env_var_secret_value.get_exposed_value());
-                } else {
-                    return Err(format!(
-                        "Invalid pattern '{}' encountered in string: {}",
-                        ctpl.get(0).unwrap().as_str(),
-                        s
-                    )
-                    .int_err()
-                    .into());
-                }
-            } else {
-                if let std::borrow::Cow::Owned(_) = &s {
-                    tracing::debug!(%s, "String after template substitution");
-                }
-                return Ok(s);
-            }
-        }
+        let lookup_fn = |env_name: &str| -> Result<String, PollingIngestError> {
+            let value = self
+                .dataset_key_value_svc
+                .find_dataset_env_var_value_by_key(env_name, dataset_env_vars)
+                .int_err()?;
+            let exposed_value = value.get_exposed_value();
+            Ok(exposed_value.to_string())
+        };
```

Comment:
What happens, if variable is not defined? Making an internal error in this case might be too rough.
We would rather prefer printing a valuable error message, so think of what errors makes sense.
Its definitely not a `PollingIngestError`, it is a too wide kind for your situation.